urlbd77b745af_KokaKiwi_MCLauncher_tgz-pJ8-net_contrapunctus_lzma_BenchmarkJ8/src/com/kokakiwi/mclauncher/core/updater/FileExtractor.java

```diff
@@ -0,0 +1,218 @@
+            {
+                extractZIP(api, source, file, dest, min, max);
+                pass = true;
+            }
+            
+            if (ext.equalsIgnoreCase("jar"))
+            {
+                extractJAR(api, source, file, dest, min, max);
+                pass = true;
+            }
+            
+            if (!recursive || pass)
+            {
+                i = -1;
+            }
+        }
+        
+        return false;
+    }
+    
+    public static File extractLZMA(LauncherAPI api, GameFile source, File file,
+            File dest, int min, int max) throws Exception
+    {
+        final File uncompressed = new File(dest, file.getName().substring(0,
+                file.getName().lastIndexOf('.')));
+        
+        final FileInputStream fileInput = new FileInputStream(file);
+        
+        InputStream in = new LzmaInputStream(fileInput);
+        
+        OutputStream out = new FileOutputStream(uncompressed);
+        
+        final byte[] buffer = new byte[16384];
+        
+        int ret = in.read(buffer);
+        while (ret >= 1)
+        {
+            out.write(buffer, 0, ret);
+            ret = in.read(buffer);
+        }
+        
+        in.close();
+        out.close();
+        
+        in = null;
+        out = null;
+        
+        return uncompressed;
+    }
+    
+    @SuppressWarnings("unchecked")
+    public static void extractZIP(LauncherAPI api, GameFile source, File file,
+            File dest, int min, int max) throws Exception
+    {
+        final ZipFile zip = new ZipFile(file);
+        
+        Enumeration<ZipEntry> entries = (Enumeration<ZipEntry>) zip.entries();
+        
+        int total = 0;
+        while (entries.hasMoreElements())
+        {
+            final ZipEntry entry = entries.nextElement();
+            
+            if (!entry.isDirectory())
+            {
+                total += entry.getSize();
+            }
+        }
+        
+        entries = (Enumeration<ZipEntry>) zip.entries();
+        final int current = 0;
+        
+        while (entries.hasMoreElements())
+        {
+            final ZipEntry entry = entries.nextElement();
+            
+            if (entry.isDirectory())
+            {
+                final File dir = new File(dest, entry.getName());
+                if (!dir.exists())
+                {
+                    dir.mkdirs();
+                }
+            }
+            else
+            {
+                final File e = new File(dest, entry.getName());
+                copyStream(api, source, zip.getInputStream(entry),
+                        new FileOutputStream(e), current, total, min, max);
+            }
+        }
+        
+        zip.close();
+    }
+    
+    public static void extractJAR(LauncherAPI api, GameFile source, File file,
+            File dest, int min, int max) throws Exception
+    {
+        JarFile jar = null;
+        try
+        {
+            jar = new JarFile(file);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *jar*. Execution paths that do not contain closure statements were detected. Ensure *jar* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).